File: Data Structures & Algorithms/products-of-array-discluding-self/submission-1.cpp
class Solution {
public:
    vector<int> productExceptSelf(vector<int>& nums) {
        int n = nums.size();
        vector<int> out(n, 1);

        for (int i = 1; i < n; ++i) {
            out[i] = out[i - 1] * nums[i - 1];
        }

        int postfix = 1;
        for (int i = n - 1; i > -1; --i) {
            out[i] *= postfix;
            postfix *= nums[i];
        }

        return out;

    }
};

File: Data Structures & Algorithms/top-k-elements-in-list/submission-1.py
from collections import defaultdict

class Solution:
    def topKFrequent(self, nums: List[int], k: int) -> List[int]:
        
        ## sorting solution:
        # frequency_table = defaultdict(int)
        # for num in nums:
        #     frequency_table[num] += 1        
        # result = []
        # sorted_values = sorted(frequency_table.values(), reverse = True)[:k]
        # for key, value in frequency_table.items():
        #     if value in sorted_values:
        #         result.append(key)
        # return result

        # bucket sort
        frequency_map = {}
        for num in nums:
            frequency_map[num] = 1 + frequency_map.get(num,0)

        frequency_array = [[] for i in range(len(nums) + 1)]

        for key, value in frequency_map.items():
            frequency_array[value].append(key)

        res = []
        for i in range(len(frequency_array) - 1, 0, -1):
            for n in frequency_array[i]:
                res.append(n)
                if len(res) == k:
                    return res

            



        